cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java

```diff
@@ -298,4 +314,96 @@ public MethodVisitor visitMethod(int access, String name, String desc, String si
+    if (!needRewrite) {
+      return findClass(name);
+    }
+
+    ClassReader cr = new ClassReader(createConfArchive);
+    ClassWriter cw = new ClassWriter(ClassWriter.COMPUTE_MAXS);
+    cr.accept(new ClassVisitor(Opcodes.ASM5, cw) {
+      @Override
+      public MethodVisitor visitMethod(final int access, final String name,
+                                       final String desc, String signature, String[] exceptions) {
+        MethodVisitor mv = super.visitMethod(access, name, desc, signature, exceptions);
+
+        // Only rewrite the createConfArchive method
+        if (!"createConfArchive".equals(name)) {
+          return mv;
+        }
+
+        // Check if it's a recognizable return type.
+        // Spark 1.5+ return type is File
+        boolean isReturnFile = Type.getReturnType(desc).equals(Type.getType(File.class));
+        Type optionType = Type.getObjectType("scala/Option");
+        if (!isReturnFile) {
+          // Spark 1.4 return type is Option<File>
+          if (!Type.getReturnType(desc).equals(optionType)) {
+            // Unknown type. Not going to modify the code.
+            return mv;
+          }
+        }
+
+        // Generate this for Spark 1.5+
+        // return SparkRuntimeUtils.createConfArchive(this.sparkConf, SPARK_CONF_FILE,
+        //                                            LOCALIZED_CONF_DIR, LOCALIZED_CONF_DIR_ZIP);
+        // Generate this for Spark 1.4
+        // return Option.apply(SparkRuntimeUtils.createConfArchive(this.sparkConf, SPARK_CONF_FILE,
+        //                                                         LOCALIZED_CONF_DIR, LOCALIZED_CONF_DIR_ZIP));
+        GeneratorAdapter mg = new GeneratorAdapter(mv, access, name, desc);
+
+        // load this.sparkConf to the stack
+        mg.loadThis();
+        mg.getField(Type.getObjectType("org/apache/spark/deploy/yarn/Client"), "sparkConf", SPARK_CONF_TYPE);
+
+        // push three constants to the stack
+        mg.visitLdcInsn(SPARK_CONF_FILE);
+        mg.visitLdcInsn(LOCALIZED_CONF_DIR);
+        mg.visitLdcInsn(LOCALIZED_CONF_DIR_ZIP);
+
+        // call SparkRuntimeUtils.createConfArchive, return a File and leave it in stack
+        Type stringType = Type.getType(String.class);
+        mg.invokeStatic(Type.getType(SparkRuntimeUtils.class),
+                        new Method("createConfArchive", Type.getType(File.class),
+                                   new Type[] { SPARK_CONF_TYPE, stringType, stringType, stringType}));
+        if (isReturnFile) {
+          // Spark 1.5+ return type is File, hence just return the File from the stack
+          mg.returnValue();
+          mg.endMethod();
+        } else {
+          // Spark 1.4 return type is Option<File>
+          // return Option.apply(<file from stack>);
+          // where the file is actually just popped from the stack
+          mg.invokeStatic(optionType, new Method("apply", optionType, new Type[] { Type.getType(Object.class) }));
```

Comment:
I meant why the change to not use ``Methods.getMethod``, but you answered that already